test-framework/sudo-compliance-tests/src/timestamp.rs

```diff
@@ -356,3 +359,124 @@ fn credential_cache_is_shared_between_sibling_shells() -> Result<()> {
+        ))
+        .as_user(USERNAME)
+        .exec(&env)?
+        .assert_success()
+}
+
+#[test]
+fn cached_credential_not_shared_with_target_user_that_are_not_self() -> Result<()> {
+    let second_target_user = "ghost";
+    let env = Env("ALL ALL=(ALL:ALL) ALL")
+        .user(User(USERNAME).password(PASSWORD))
+        .user(second_target_user)
+        .build()?;
+
+    let output = Command::new("sh")
+        .arg("-c")
+        .arg(format!(
+            "echo {PASSWORD} | sudo -S true; sudo -u {second_target_user} sudo -S true"
+        ))
+        .as_user(USERNAME)
+        .exec(&env)?;
+
+    assert!(!output.status().success());
+
+    assert_eq!(Some(1), output.status().code());
+
+    let diagnostic = if sudo_test::is_original_sudo() {
+        "a password is required"
+    } else {
+        "Maximum 3 incorrect authentication attempts"
+    };
+
+    assert_contains!(output.stderr(), diagnostic);
+
+    Ok(())
+}
+
+#[test]
+#[ignore = "gh388"]
+fn cached_credential_shared_with_target_user_that_is_self() -> Result<()> {
+    let env = Env(format!("{USERNAME} ALL=(ALL:ALL) ALL"))
+        .user(User(USERNAME).password(PASSWORD))
+        .build()?;
+
+    // FIXME switch back to `exec.assert_success`. this operation makes sudo-rs hang so we use
+    // `spawn` + `try_wait` polling here to avoid blocking forever
+    let mut child = Command::new("sh")
+        .arg("-c")
+        .arg(format!(
+            "echo {PASSWORD} | sudo -S true; sudo -u {USERNAME} env '{SUDO_RS_IS_UNSTABLE}' sudo true"
+        ))
+        .as_user(USERNAME)
+        .tty(true)
+        .spawn(&env)?;
+
+    for _ in 0..5 {
+        if let Some(status) = child.try_wait()? {
+            assert!(status.success());
+            return Ok(());
+        }
+
+        thread::sleep(Duration::from_secs(1));
+    }
+
+    panic!("timed out")
+}
+
+#[test]
+fn flag_reset_timestamp_plus_command_prompts_for_password() -> Result<()> {
+    let env = Env(format!("{USERNAME} ALL=(ALL:ALL) ALL"))
+        .user(User(USERNAME).password(PASSWORD))
+        .build()?;
+
+    let output = Command::new("sh")
+        .arg("-c")
+        .arg(format!("echo {PASSWORD} | sudo -S true; sudo -k true"))
+        .as_user(USERNAME)
+        .exec(&env)?;
+
+    assert!(!output.status().success());
+    assert_eq!(Some(1), output.status().code());
+
+    let diagnostic = if sudo_test::is_original_sudo() {
+        "a password is required"
+    } else {
+        "incorrect authentication attempt"
+    };
+    assert_contains!(output.stderr(), diagnostic);
+
+    Ok(())
+}
+
+#[test]
+fn flag_reset_timestamp_plus_command_does_not_invalidate_credential_cache() -> Result<()> {
+    let env = Env(format!("{USERNAME} ALL=(ALL:ALL) ALL"))
+        .user(User(USERNAME).password(PASSWORD))
+        .build()?;
+
+    Command::new("sh")
+        .arg("-c")
+        .arg(format!(
+            "echo {PASSWORD} | sudo -S true; sudo -k true; [ $? -eq 1 ] || exit 2; sudo true"
```

Comment:
This test confused me a bit. I imagine `sudo -k true` fails because no password was prompted so `[ $? -eq 1 ] || exit 2` does not call `exit`. But then shouldn't  `sudo true` fail again?

Reply:
oh ok so it only request the password again but the credentials are valid afterwards got it.